File: 0732-my-calendar-iii/0732-my-calendar-iii.cpp
class MyCalendarThree {
public:
    unordered_map<int, int> seg;
    unordered_map<int, int> lazy;
    
    MyCalendarThree() {}

    void update(int idx, int low, int high, int l, int r) {
        if (lazy[idx] != 0) {
            seg[idx] += lazy[idx];
            if (low != high) {
                lazy[2 * idx + 1] += lazy[idx];
                lazy[2 * idx + 2] += lazy[idx];
            }
            lazy[idx] = 0;
        }
        if (l > high || r < low) return;
        if (l <= low && high <= r) {
            seg[idx] += 1;
            if (low != high) {
                lazy[2 * idx + 1] += 1;
                lazy[2 * idx + 2] += 1;
            }
            return;
        }
        int mid = (low + high) >> 1;
        update(2 * idx + 1, low, mid, l, r);
        update(2 * idx + 2, mid + 1, high, l, r);
        seg[idx] = max(seg[2 * idx + 1], seg[2 * idx + 2]);
    }

    int book(int startTime, int endTime) {
        update(0, 0, 1000000000, startTime, endTime - 1);
        return seg[0];
    }
};

File: 0732-my-calendar-iii/0732-my-calendar-iii.py
class MyCalendarThree(object):
    def __init__(self):
        self.max = collections.defaultdict(int)
        self.lazy = collections.defaultdict(int)

    def update(self, l, r, left=0, right=int(1e9), id=1):
        if right <= l or r <= left: return
        if l <= left and right <= r:
            self.max[id] += 1
            self.lazy[id] += 1
        else:
            md = (left + right) >> 1
            lc, rc = id*2, id*2+1
            self.update(l, r, left, md, lc)
            self.update(l, r, md, right, rc)
            self.max[id] = self.lazy[id] + max(self.max[lc], self.max[rc])

    def book(self, start, end):
        self.update(start, end)
        return self.max[1]